File: app/src/main/java/com/regaan/lockroot/vault/PasswordGenerator.kt
package com.regaan.lockroot.vault

import java.security.SecureRandom

class PasswordGenerator(
    private val secureRandom: SecureRandom = SecureRandom(),
) {
    fun generate(
        length: Int = 24,
        lowercase: Boolean = true,
        uppercase: Boolean = true,
        numbers: Boolean = true,
        symbols: Boolean = true,
    ): String {
        require(length in 12..128) { "Password length must be between 12 and 128." }

        val groups = buildList {
            if (lowercase) add(LOWERCASE)
            if (uppercase) add(UPPERCASE)
            if (numbers) add(NUMBERS)
            if (symbols) add(SYMBOLS)
        }
        require(groups.isNotEmpty()) { "At least one character group is required." }
        require(length >= groups.size) { "Length must fit all selected groups." }

        val result = mutableListOf<Char>()
        groups.forEach { result.add(randomChar(it)) }

        val allChars = groups.joinToString("")
        while (result.size < length) {
            result.add(randomChar(allChars))
        }

        for (index in result.indices.reversed()) {
            val swapIndex = secureRandom.nextInt(index + 1)
            val tmp = result[index]
            result[index] = result[swapIndex]
            result[swapIndex] = tmp
        }

        return result.joinToString("")
    }

    private fun randomChar(chars: String): Char {
        val bound = chars.length
        val max = Int.MAX_VALUE - (Int.MAX_VALUE % bound)
        var value: Int
        do {
            value = secureRandom.nextInt() and Int.MAX_VALUE
        } while (value >= max)
        return chars[value % bound]
    }

    companion object {
        private const val LOWERCASE = "abcdefghijkmnopqrstuvwxyz"
        private const val UPPERCASE = "ABCDEFGHJKLMNPQRSTUVWXYZ"
        private const val NUMBERS = "23456789"
        private const val SYMBOLS = "!@#$%^&*()-_=+[]{};:,.?"
    }
}

File: macos/LockrootMac/Sources/LockrootMac/Services/PasswordGenerator.swift
import Foundation
import Security

final class PasswordGenerator {
    private static let lowercase = Array("abcdefghijkmnopqrstuvwxyz")
    private static let uppercase = Array("ABCDEFGHJKLMNPQRSTUVWXYZ")
    private static let numbers = Array("23456789")
    private static let symbols = Array("!@#$%^&*()-_=+[]{};:,.?")

    func generate(
        length: Int = 24,
        lowercase: Bool = true,
        uppercase: Bool = true,
        numbers: Bool = true,
        symbols: Bool = true
    ) throws -> String {
        guard (12...128).contains(length) else {
            throw CryptoError.unsupportedFormat("Password length must be between 12 and 128.")
        }

        var groups: [[Character]] = []
        if lowercase { groups.append(Self.lowercase) }
        if uppercase { groups.append(Self.uppercase) }
        if numbers { groups.append(Self.numbers) }
        if symbols { groups.append(Self.symbols) }
        guard !groups.isEmpty else {
            throw CryptoError.unsupportedFormat("At least one character group is required.")
        }
        guard length >= groups.count else {
            throw CryptoError.unsupportedFormat("Length must fit all selected groups.")
        }

        var result = try groups.map { try randomCharacter(from: $0) }
        let all = groups.flatMap { $0 }
        while result.count < length {
            result.append(try randomCharacter(from: all))
        }

        for index in result.indices.reversed() {
            let swapIndex = try randomInt(upperBound: index + 1)
            result.swapAt(index, swapIndex)
        }
        return String(result)
    }

    private func randomCharacter(from characters: [Character]) throws -> Character {
        characters[try randomInt(upperBound: characters.count)]
    }

    private func randomInt(upperBound: Int) throws -> Int {
        precondition(upperBound > 0)
        let max = UInt32.max - (UInt32.max % UInt32(upperBound))
        var value: UInt32 = 0
        repeat {
            var bytes = [UInt8](repeating: 0, count: MemoryLayout<UInt32>.size)
            let count = bytes.count
            let status = bytes.withUnsafeMutableBytes { raw in
                guard let baseAddress = raw.baseAddress else { return errSecParam }
                return SecRandomCopyBytes(kSecRandomDefault, count, baseAddress)
            }
            guard status == errSecSuccess else { throw CryptoError.randomFailure }
            value =
                UInt32(bytes[0]) |
                (UInt32(bytes[1]) << 8) |
                (UInt32(bytes[2]) << 16) |
                (UInt32(bytes[3]) << 24)
        } while value >= max
        return Int(value % UInt32(upperBound))
    }
}
